File: tests/level-reading-test.c
#include <stdio.h>
#include <gst/gst.h>
#include <stdbool.h>

static GMainLoop *loop;

static gboolean bus_call(GstBus *bus, GstMessage *msg, void *user_data)
{
	switch(GST_MESSAGE_TYPE(msg)) {
		case GST_MESSAGE_EOS: {
			//g_message("Dziesmas beigas.");
			g_main_loop_quit(loop);
			break;
		}
		case GST_MESSAGE_ELEMENT: {
			
			const GstStructure *st = gst_message_get_structure(msg);
			const gchar *name = gst_structure_get_name(st);
			printf("%s\n", name);
			// if not equal, break, it's not level element message
			if(strcmp(name, "level") != 0)
				break;
			
			// get value list of channel median powers
			const GValue *list_value = gst_structure_get_value(st, "rms");
			GValueArray *rms_list = (GValueArray *) g_value_get_boxed(list_value);
			const GValue *rms_value;
			
			guint channels = rms_list->n_values;
			printf("Channels: %i\n",channels);
			
			int i;
			for(i=0;i<channels;i++) {
				rms_value = g_value_array_get_nth(rms_list, i);
				printf("RMS #%i: %f\n", i+1, g_value_get_double(rms_value));
			}
		}
		default:
			break;
	}
}

int main(int argc, char *argv[]) {
// This is main
	gst_init(&argc, &argv);
	
	// Create main loop
	loop = g_main_loop_new(NULL, FALSE);
	
	GstElement *pipeline;
	GstBus *bus;
	gchar location[] = "test.flac";
	
	pipeline = gst_parse_launch("filesrc name=src ! decodebin ! audioconvert ! level message=true name=level_element ! fakesink", NULL);
	//pipeline = gst_element_factory_make("playbin", "player");
	
	// set up object values
	GstElement *filesrc = gst_bin_get_by_name(GST_BIN(pipeline), "src");
	g_object_set(G_OBJECT(filesrc), "location", location, NULL);
	gst_object_unref(filesrc);
	
	bus = gst_pipeline_get_bus(GST_PIPELINE(pipeline));
	gst_bus_add_watch(bus, bus_call, NULL);
	gst_object_unref(bus);
	
	gst_element_set_state(GST_ELEMENT(pipeline), GST_STATE_PLAYING);
	
	g_main_loop_run(loop);
	
	gst_element_set_state(GST_ELEMENT(pipeline), GST_STATE_NULL);
	
	gst_object_unref(GST_OBJECT(pipeline));
}

File: tests/level-reading-test.py
# test level reading execution using Python/GObject/GI
# set location property of filesrc element to file you want to read
# time python level-reading-test.py > log.log
# This will measure time spent to read this file's levels
#!/usr/bin/python

from gi.repository import Gst, Gtk

def bus_element(bus, message):
	st = message.get_structure()
	if st.get_name() == "level":
		levels = st.get_value("rms")
		loadingLength = st.get_value("endtime")
		channel = 1
		for level in levels:
			print("Channel #" + str(channel) + " " + str(level) + " " + str(loadingLength))
			channel += 1

def bus_eos(bus, message):
	global loadingPipeline
	loadingPipeline.set_state(Gst.State.NULL)
	Gtk.main_quit()
	return

Gst.init(None)
#LEVEL_INTERVAL = 0.1

pipe = """filesrc name=src ! decodebin ! audioconvert ! level message=true name=level_element ! fakesink"""
loadingPipeline = Gst.parse_launch(pipe)

filesrc = loadingPipeline.get_by_name("src")
#level = loadingPipeline.get_by_name("level_element")

filesrc.set_property("location", "test.flac")
#level.set_property("interval", int(LEVEL_INTERVAL * Gst.SECOND))

bus = loadingPipeline.get_bus()
bus.add_signal_watch()
bus.connect("message::eos", bus_eos)
bus.connect("message::element", bus_element)
loadingPipeline.set_state(Gst.State.PLAYING)

Gtk.main()
